crates/sui-json-rpc/src/error.rs

```diff
@@ -95,19 +94,99 @@ impl Error {
+                    );
+                    RpcError::Call(CallError::Custom(error_object))
+                }
+                QuorumDriverError::TimeoutBeforeFinality
+                | QuorumDriverError::FailedWithTransientErrorAfterMaximumAttempts { .. } => {
+                    let error_object =
+                        ErrorObject::owned(TRANSIENT_ERROR_CODE, err.to_string(), None::<()>);
+                    RpcError::Call(CallError::Custom(error_object))
+                }
+                QuorumDriverError::ObjectsDoubleUsed {
+                    conflicting_txes,
+                    retried_tx,
+                    retried_tx_success,
+                } => {
+                    let error_message = format!(
+                        "Failed to sign transaction by a quorum of validators because of locked objects. Retried a conflicting transaction {:?}, success: {:?}",
+                        retried_tx,
+                        retried_tx_success
+                    );
+
+                    let mut new_map: BTreeMap<TransactionDigest, Vec<ObjectRef>> = BTreeMap::new();
+
+                    for (digest, (pairs, _)) in conflicting_txes {
+                        let mut new_vec = Vec::new();
+
+                        for (_authority, obj_ref) in pairs {
+                            new_vec.push(obj_ref);
+                        }
+
+                        new_map.insert(digest, new_vec);
+                    }
+
+                    let error_object =
+                        ErrorObject::owned(INVALID_PARAMS_CODE, error_message, Some(new_map));
+                    RpcError::Call(CallError::Custom(error_object))
+                }
                 QuorumDriverError::NonRecoverableTransactionError { errors } => {
-                    // Note: we probably want a more precise error than `INVALID_PARAMS_CODE`
-                    // but to keep the error code consistent we still use `INVALID_PARAMS_CODE`
+                    let new_errors: Vec<String> = errors
+                        .into_iter()
+                        .filter(|(err, _, _)| !err.is_retryable().0) // consider retryable errors as transient errors
+                        .map(|(err, _, _)| {
+                            match err {
+                                // Use inner UserInputError's Display instead of SuiError::UserInputError, which renders UserInputError in debug format
+                                SuiError::UserInputError { error } => error.to_string(),
+                                _ => err.to_string(),
+                            }
+                        })
+                        .collect();
+
+                    let (error_code, error_msg) = if new_errors.is_empty() {
+                        (
+                            TRANSIENT_ERROR_CODE,
+                            "Transaction execution failed due to transient errors. Please try again.",
+                        )
+                    } else {
```

Comment:
this is not supposed to happen when TransactionOrchestrator/QuorumDriver returns `NonRecoverableTransactionError`. So we can safely assert here to make the code simpler

Reply:
L147 onwards filters out retryable SuiErrors, such as RpcError. I think we may end up with an empty vector if the suierrors all end up being retryable

Reply:
`NonRecoverableTransactionError` should have at least 1/3 NonRecoverable errors. So if we filter retryable errors in line 147 and get an empty vector, it indicates a bug somewhere

Reply:
Ahhh yes I see, totally went over my head. Thanks! Makes sense